### PR TITLE
Implement cast between Date and Timestamp

### LINKED ISSUE #4644: Comparing a `Timestamp` to a `Date32` fails
**Describe the bug**
Comparing a `Timestamp` to a `Date32` fails.

**To Reproduce**
Add the following test to [`timestamps.slt`](https://github.com/apache/arrow-datafusion/blob/5c558e9f6b3fe29ef0d5f78ec9465852d020d989/datafusion/core/tests/sqllogictests/test_files/timestamps.slt#L3):

```
# Test that we can compare a timestamp to a string casted to a date
query C rowsort
select * from foo where ts > '2000-01-01'::date;
----
2 2000-02-01T00:00:00
3 2000-03-01T00:00:00
```

This fails w/:

```
[timestamps.slt] Running query: "select * from foo where ts > '2000-01-01'::date;"
Error: SqlLogicTest(query failed: DataFusion error: Internal error: The type of Timestamp(Nanosecond, None) > Date32 of binary physical should be same. This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker
```

**Expected behavior**
It works.

**Additional context**
Tested on https://github.com/apache/arrow-datafusion/commit/5c558e9f6b3fe29ef0d5f78ec9465852d020d989 .

This is likely due to a like in [`temporal_coercion`](https://github.com/apache/arrow-datafusion/blob/5c558e9f6b3fe29ef0d5f78ec9465852d020d989/datafusion/expr/src/type_coercion/binary.rs#L539-L539) as well as this workaround:

https://github.com/apache/arrow-datafusion/blob/5c558e9f6b3fe29ef0d5f78ec9465852d020d989/datafusion/optimizer/src/type_coercion.rs#L232-L239

Also see #3419.

### LINKED ISSUE #4672: Timestamp <-> Date32 compare doesn't work
**Describe the bug**
Make conversion seemless

**To Reproduce**
```
❯ select now() = '2000-01-01'::date;
Plan("'Timestamp(Nanosecond, Some(\"+00:00\")) = Date32' can't be evaluated because there isn't a common type to coerce the types to")
```
**Expected behavior**
Shouldnt fail

**Additional context**

### PR DESCRIPTION
**Pull request by comphead**
# Which issue does this PR close?

Closes #4644 
Closes #4672.

# Rationale for this change
Add cast between Date and Timestamp

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/expr/src/type_coercion/binary.rs [hunk 1]
```diff
@@ -575,6 +575,8 @@ fn temporal_coercion(lhs_type: &DataType, rhs_type: &DataType) -> Option<DataTyp
         },
         (Timestamp(_, tz), Utf8) => Some(Timestamp(TimeUnit::Nanosecond, tz.clone())),
         (Utf8, Timestamp(_, tz)) => Some(Timestamp(TimeUnit::Nanosecond, tz.clone())),
+        (Timestamp(_, _), Date32) => Some(Date32),
```
Comment (alamb) [inline datafusion/expr/src/type_coercion/binary.rs:578]: 👍

**Review by alamb — APPROVED** [review]
LGTM -- thank you @comphead

**Comment by alamb** [comment]
Thanks again @comphead

**Comment by liukun4515** [comment]
Sorry for the late reply, I am confused about the implementation for comparing between timestamp and date type.

We have moved the `type coercion` logic to the logical phase in the `type_coercion` rule, but this feature has side affect to this.

@alamb @comphead

**Comment by alamb** [comment]
> Sorry for the late reply, I am confused about the implementation for comparing between timestamp and date type.

It is an excellent point @liukun4515  -- it turns out that coercion is not necessary -- see https://github.com/apache/arrow-datafusion/pull/4745

**Comment by alamb** [comment]
Actually, it appears @liukun4515  already had a PR to do this 🤦 

https://github.com/apache/arrow-datafusion/pull/4741

**Pull request merged**